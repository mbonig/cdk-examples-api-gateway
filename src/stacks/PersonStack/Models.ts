import { IdentifiableModel } from '../common/handlers/IdentifiableModel';

export interface Person extends IdentifiableModel {
  firstName: string;
  lastName: string;
}

