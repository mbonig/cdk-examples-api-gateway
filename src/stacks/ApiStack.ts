import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Deployment, RestApi, Stage } from 'aws-cdk-lib/aws-apigateway';
import { Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
import { AttributeType, TableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { ARecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { ApiGateway } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

export interface ApiStackProps extends StackProps {
  hostedZoneId: string;
}

export class ApiStack extends Stack {
  api: RestApi;
  table: TableV2;
  private deployment!: Deployment;
  private apiDomainName: string;

  constructor(scope: Construct, id: string, private props: ApiStackProps) {
    super(scope, id, props);

    const apiDomainName = this.apiDomainName = 'api.example.matthewbonig.com';
    this.api = this.createApi();
    this.table = this.createTable();

    new CfnOutput(this, 'BaseApiGatewayURL', {
      value: `https://${this.api.restApiId}.execute-api.${this.region}.amazonaws.com/prod/`,
    });
    new CfnOutput(this, 'CustomURL', {
      value: `https://${apiDomainName}/v1/`,
    });

  }

  private createApi(): RestApi {
    const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: this.props.hostedZoneId,
      zoneName: 'example.matthewbonig.com',
    });
    const api = new RestApi(this, 'MyRestApi', {
      deploy: false,
      restApiName: 'CdkCourseApi',
      domainName: {
        domainName: this.apiDomainName,
        certificate: new Certificate(this, 'Certificate', {
          domainName: this.apiDomainName,
          validation: CertificateValidation.fromDns(hostedZone),
        }),
      },
    });

    new ARecord(this, 'ApiAliasRecord', {
      target: {
        aliasTarget: new ApiGateway(api),
      },
      recordName: 'api',
      zone: hostedZone,
    });

    api.root.addMethod('ANY');

    const deployment = this.deployment = new Deployment(this, 'Deployment', {
      api: api,
    });
    new Stage(this, 'Prod', {
      deployment,
      stageName: 'v1',
    });

    return api;
  }

  private createTable(): TableV2 {
    return new TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'pk',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });
  }

  public addToDeployment(resources: any[]) {
    for (const resource of resources) {
      this.deployment.addToLogicalId(resource);
    }
  }
}
