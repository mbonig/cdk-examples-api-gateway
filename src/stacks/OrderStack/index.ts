export * from './OrderStack';
export { OrderStackTests, OrderStackTestProps } from './OrderStackTests/OrderStackTests';

