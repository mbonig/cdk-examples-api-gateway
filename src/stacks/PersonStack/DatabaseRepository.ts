import { IdentifiableModel } from '../common/handlers/IdentifiableModel';

export interface IDatabaseRepository<ModelType extends IdentifiableModel> {
  get(id: string): Promise<ModelType>;

  getAll(): Promise<ModelType[]>;

  save(data: ModelType): Promise<ModelType>;
}
