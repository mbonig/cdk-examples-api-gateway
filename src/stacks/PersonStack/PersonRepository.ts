import { IDatabaseRepository } from './DatabaseRepository';
import { Person } from './Models';
import { IDatabaseAdapter, TableModel } from '../common/IDatabaseAdapter';


export interface IPersonRepository extends IDatabaseRepository<Person> {

}

export class PersonRepository implements IPersonRepository {
  constructor(private databaseAdapter: IDatabaseAdapter, private indexName: string) {

  }

  async getAll(): Promise<Person[]> {
    let people = await this.databaseAdapter.getAll(this.indexName);

    return people.map((person: TableModel) => ({
      id: person.sk,
      firstName: person.firstName,
      lastName: person.lastName,
    }));
  }

  async get(id: string): Promise<Person> {
    const person = await this.databaseAdapter.getItem(id, id);
    return { personId: person.sk, firstName: person.firstName, lastName: person.lastName };
  }

  async save(data: Person): Promise<Person> {
    const person: Person = {
      ...data,
    };
    await this.databaseAdapter.putItem({ pk: 'person', sk: person.personId, ...person });
    return person;
  }
}

