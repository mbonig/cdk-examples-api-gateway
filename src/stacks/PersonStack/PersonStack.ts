import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { INDEX_NAME, TABLE_NAME } from './PersonStack.GetPersonHandler';
import { RestApiResourceStack } from '../RestApiResourceStack';
import { RestResourceProps } from '../RestResourceStackProps';

export interface PersonStackProps extends RestResourceProps {
}

export class PersonStack extends RestApiResourceStack {

  public deploymentDependencies: any[] = [];

  constructor(scope: Construct, id: string, props: PersonStackProps) {
    super(scope, id, props);

    const indexName = 'people';
    const getHandler = new NodejsFunction(this, 'GetPersonHandler', {
      environment: {
        [TABLE_NAME]: props.table.tableName,
        [INDEX_NAME]: indexName,
      },
    });
    props.table.grantReadData(getHandler);

    const personResource = this.api.root.addResource('person', {});

    const getMethod = personResource.addMethod('GET', new LambdaIntegration(getHandler));

    const getByIdMethod = personResource
      .addResource('{id}', {})
      .addMethod('GET', new LambdaIntegration(getHandler));

    props.table.addGlobalSecondaryIndex({
      indexName: indexName,
      partitionKey: {
        name: 'personId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });

    this.deploymentDependencies.push({ method: getMethod.methodId }, { method: getByIdMethod.methodId });
  }

}
