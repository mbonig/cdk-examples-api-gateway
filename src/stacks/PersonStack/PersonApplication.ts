import { Person } from './Models';
import { IPersonRepository } from './PersonRepository';

export interface IPersonApplication {
  getById(id: string): Promise<Person>;

  getAll(): Promise<Person[]>;

}

export class PersonApplication implements IPersonApplication {
  constructor(private personRepository: IPersonRepository) {
  }


  async getById(id: string): Promise<Person> {
    if (!id) {
      throw new Error('Please provide an id');
    }
    return this.personRepository.get(id);
  }

  async getAll(): Promise<Person[]> {
    return this.personRepository.getAll();
  }

}
