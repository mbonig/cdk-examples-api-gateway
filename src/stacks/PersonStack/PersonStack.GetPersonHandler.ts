import { APIGatewayEvent } from 'aws-lambda';
import { PersonApplication } from './PersonApplication';
import { PersonRepository } from './PersonRepository';
import { DynamoDBAdapter } from '../common/IDatabaseAdapter';

export const TABLE_NAME = 'TABLE_NAME';
export const INDEX_NAME = 'INDEX_NAME';

export const handler = async (event: APIGatewayEvent) => {
  console.log(JSON.stringify(event, null, 2));
  const indexName = process.env[INDEX_NAME]!;
  const personApplication = new PersonApplication(
    new PersonRepository(
      new DynamoDBAdapter(process.env[TABLE_NAME]!),
      indexName,
    ),
  );

  if (event.pathParameters?.id) {
    const id = event.pathParameters.id;
    const person = await personApplication.getById(id);
    if (!person) {
      return {
        statusCode: 404,
        body: JSON.stringify({ message: 'Person not found' }),
      };
    }
    return {
      statusCode: 200,
      body: JSON.stringify(person),
    };
  }

  const persons = await personApplication.getAll();
  return {
    statusCode: 200,
    body: JSON.stringify(persons),
  };
};
