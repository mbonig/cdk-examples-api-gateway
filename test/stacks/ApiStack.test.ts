import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { ApiStack } from '../../src/stacks/ApiStack';
import { PersonStack } from '../../src/stacks/PersonStack';

test('Snapshot', () => {
  const app = new App();
  const stack = new ApiStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

test('Snapshot with added dependencies', () => {
  const app = new App();
  const stack = new ApiStack(app, 'test');
  const personStack = new PersonStack(stack, 'Person', { api: stack.api, table: stack.table });
  stack.addToDeployment(personStack.deploymentDependencies);
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
