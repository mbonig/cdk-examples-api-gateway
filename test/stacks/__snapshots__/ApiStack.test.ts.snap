// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "BaseApiGatewayURL": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "MyRestApi2D1F47A9",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".amazonaws.com/prod/",
          ],
        ],
      },
    },
    "CustomURL": {
      "Value": "https://api.example.matthewbonig.com/v1/",
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiAliasRecordBCE933A3": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "MyRestApiCustomDomain3EE86491",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "MyRestApiCustomDomain3EE86491",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "Z008193231WFP7BBYULYK",
        "Name": "api.example.matthewbonig.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "api.example.matthewbonig.com",
        "DomainValidationOptions": [
          {
            "DomainName": "api.example.matthewbonig.com",
            "HostedZoneId": "Z008193231WFP7BBYULYK",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/Certificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "Deployment37BBD5E403702306056bac8e3ef6cec63ad8a852": {
      "DependsOn": [
        "MyRestApiANY05143F93",
      ],
      "Properties": {
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MyRestApi2D1F47A9": {
      "Properties": {
        "Name": "CdkCourseApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MyRestApiANY05143F93": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "Type": "MOCK",
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "MyRestApi2D1F47A9",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MyRestApiAccount2FB6DB7A": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "MyRestApi2D1F47A9",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "MyRestApiCloudWatchRoleD4042E8E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "MyRestApiCloudWatchRoleD4042E8E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "MyRestApiCustomDomain3EE86491": {
      "Properties": {
        "DomainName": "api.example.matthewbonig.com",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Ref": "Certificate4E7ABB08",
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "MyRestApiCustomDomainMaptestMyRestApi2A9F0BDD4FD6466B": {
      "Properties": {
        "DomainName": {
          "Ref": "MyRestApiCustomDomain3EE86491",
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "MyTable794EDED1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
    "Prod7A5F3219": {
      "DependsOn": [
        "MyRestApiAccount2FB6DB7A",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "Deployment37BBD5E403702306056bac8e3ef6cec63ad8a852",
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
        "StageName": "v1",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Snapshot with added dependencies 1`] = `
{
  "Outputs": {
    "BaseApiGatewayURL": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "MyRestApi2D1F47A9",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".amazonaws.com/prod/",
          ],
        ],
      },
    },
    "CustomURL": {
      "Value": "https://api.example.matthewbonig.com/v1/",
    },
    "ExportsOutputFnGetAttMyRestApi2D1F47A9RootResourceIdA8A99F67": {
      "Export": {
        "Name": "test:ExportsOutputFnGetAttMyRestApi2D1F47A9RootResourceIdA8A99F67",
      },
      "Value": {
        "Fn::GetAtt": [
          "MyRestApi2D1F47A9",
          "RootResourceId",
        ],
      },
    },
    "ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254": {
      "Export": {
        "Name": "test:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
      },
      "Value": {
        "Fn::GetAtt": [
          "MyTable794EDED1",
          "Arn",
        ],
      },
    },
    "ExportsOutputRefMyRestApi2D1F47A95DC88B97": {
      "Export": {
        "Name": "test:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
      },
      "Value": {
        "Ref": "MyRestApi2D1F47A9",
      },
    },
    "ExportsOutputRefMyTable794EDED1558A264E": {
      "Export": {
        "Name": "test:ExportsOutputRefMyTable794EDED1558A264E",
      },
      "Value": {
        "Ref": "MyTable794EDED1",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiAliasRecordBCE933A3": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "MyRestApiCustomDomain3EE86491",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "MyRestApiCustomDomain3EE86491",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "Z008193231WFP7BBYULYK",
        "Name": "api.example.matthewbonig.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "api.example.matthewbonig.com",
        "DomainValidationOptions": [
          {
            "DomainName": "api.example.matthewbonig.com",
            "HostedZoneId": "Z008193231WFP7BBYULYK",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/Certificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "Deployment37BBD5E45525ea0a01ac4155b999960e3c29d027": {
      "DependsOn": [
        "MyRestApiANY05143F93",
      ],
      "Properties": {
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MyRestApi2D1F47A9": {
      "Properties": {
        "Name": "CdkCourseApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MyRestApiANY05143F93": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "Type": "MOCK",
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "MyRestApi2D1F47A9",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MyRestApiAccount2FB6DB7A": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "MyRestApi2D1F47A9",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "MyRestApiCloudWatchRoleD4042E8E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "MyRestApiCloudWatchRoleD4042E8E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "MyRestApiCustomDomain3EE86491": {
      "Properties": {
        "DomainName": "api.example.matthewbonig.com",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Ref": "Certificate4E7ABB08",
        },
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "MyRestApiCustomDomainMaptestMyRestApi2A9F0BDD4FD6466B": {
      "Properties": {
        "DomainName": {
          "Ref": "MyRestApiCustomDomain3EE86491",
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "MyTable794EDED1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
          {
            "AttributeName": "personId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "people",
            "KeySchema": [
              {
                "AttributeName": "personId",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "sk",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "GlobalSecondaryIndexes": [
              {
                "IndexName": "people",
              },
            ],
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
    "Prod7A5F3219": {
      "DependsOn": [
        "MyRestApiAccount2FB6DB7A",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "Deployment37BBD5E45525ea0a01ac4155b999960e3c29d027",
        },
        "RestApiId": {
          "Ref": "MyRestApi2D1F47A9",
        },
        "StageName": "v1",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
